staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go

```diff
@@ -0,0 +1,575 @@
+		parsed := &apidiscoveryv2beta1.APIGroupDiscoveryList{}
+		if err := runtime.DecodeInto(scheme.Codecs.UniversalDecoder(), writer.data, parsed); err != nil {
+			return nil, err
+		}
+		klog.Infof("DiscoveryManager: Successfully downloaded discovery for %s", info.service.String())
+
+		// Convert discovery info into a map for convenient lookup later
+		discoMap := map[metav1.GroupVersion]apidiscoveryv2beta1.APIVersionDiscovery{}
+		for _, g := range parsed.Items {
+			for _, v := range g.Versions {
+				discoMap[metav1.GroupVersion{Group: g.Name, Version: v.Version}] = v
+			}
+		}
+
+		// Save cached result
+		cached = cachedResult{
+			discovery:   discoMap,
+			etag:        writer.Header().Get("Etag"),
+			lastUpdated: now,
+		}
+		dm.setCacheEntryForService(info.service, cached)
+		return &cached, nil
+
+	default:
+		// Unhandled response. Mark information as stale.
+		// Try again later.
+		//!TODO: After a few tries, just wipe it out?
+		// 	or after certain time?
+		cached = cachedResult{
+			discovery:   cached.discovery,
+			etag:        cached.etag,
+			lastUpdated: now,
+			stale:       true,
+		}
+		dm.setCacheEntryForService(info.service, cached)
+
+		klog.Infof("DiscoveryManager: Failed to download discovery for %v: %v %s",
+			info.service.String(), writer.respCode, writer.data)
+		return nil, fmt.Errorf("service %s returned non-success response code: %v",
+			info.service.String(), writer.respCode)
+	}
+}
+
+// Try to sync a single APIService.
+func (dm *discoveryManager) syncAPIService(apiServiceName string) error {
+	info, exists := dm.getInfoForAPIService(apiServiceName)
+
+	gv := helper.APIServiceNameToGroupVersion(apiServiceName)
+	mgv := metav1.GroupVersion{Group: gv.Group, Version: gv.Version}
+
+	if !exists {
+		// apiservice was removed. remove it from merged discovery
+		dm.mergedDiscoveryHandler.RemoveGroupVersion(mgv)
+		return nil
+	}
+
+	// Lookup last cached result for this apiservice's service.
+	now := time.Now()
+	cached, err := dm.fetchFreshDiscoveryForService(mgv, info)
+
+	info.lastReconciled = now
+	dm.setInfoForAPIService(apiServiceName, &info)
+
+	if err != nil {
+		// There was an error fetching discovery for this APIService.
+		// Just use empty GV to mark that GV exists, but no resources.
+		//
+		// TODO: Maybe also stick in a status for the version the error?
+		dm.mergedDiscoveryHandler.AddGroupVersion(gv.Group, apidiscoveryv2beta1.APIVersionDiscovery{
+			Version: gv.Version,
+		})
+		return err
+	}
+
+	// Check cache entry for this groupversion and insert it into the
+	// document if present
+	if entry, exists := cached.discovery[mgv]; exists {
+		if cached.stale {
+			entry.Freshness = apidiscoveryv2beta1.DiscoveryFreshnessStale
+		}
+		dm.mergedDiscoveryHandler.AddGroupVersion(gv.Group, entry)
+	} else {
+		// Use empty GV, since there is an APIService for it
+		//!TODO: mark with staleness of cache
+		dm.mergedDiscoveryHandler.AddGroupVersion(gv.Group, apidiscoveryv2beta1.APIVersionDiscovery{
```

Comment:
how does this happen?

Reply:
If there were errors during other parts of the fetch process (decoding, validation gvs, etc), we will not create a cache entry for the apiservice.